In [18]:
from Bio import SeqIO
import pandas as pd

In [4]:
fastaseq = SeqIO.parse("./aligned.fasta", "fasta") #sequences ist die bereits die alignierte subsampled file

In [6]:
inf = []
seq = []
seqlen = []

for seq_record in fastaseq:
   # print(seq_record.id)
    inf.append(seq_record.id)
  #  print(repr(seq_record.seq))
    seq.append(seq_record.seq)
  #  print(len(seq_record),"\n")
    seqlen.append(len(seq_record))
    

In [17]:
inf

['OV133415',
 'KEN/C120214/2022',
 'KEN/C120264/2022',
 'OX346176',
 'GMB/5229/2021',
 'KEN/SS11213/2022',
 'KEN/C120243/2022',
 'KEN/C120465/2022',
 'THA/CONI-3893/2022',
 'Switzerland/ZH-UZH-IMV-3ba67325/2022',
 'Japan/SZ-NIG-Y221912/2022',
 'CHN/20221027_114/2022',
 'KEN/SS11063/2022',
 'ZAF/NHLS-UCT-PA-J102/2022',
 'England/LSPA-325F56FF/2022',
 'ZAF/NHLS-UCT-PA-J026/2022',
 'USA/LA-BIE-LSUH003675/2023',
 'OQ520225',
 'LC655378',
 'Switzerland/GE-HUG-34933954/2021',
 'OX433662',
 'USA/WV134884/2022',
 'BHR/890702063/2023',
 'BHR/650046102/2023',
 'USA/WV090078/2021',
 'England/BRBR-326CCAB8/2023',
 'ZAF/NHLS-UCT-LA-ZB67/2022',
 'BHR/22920496163/2022',
 'OQ520232',
 'BHR/850911559/2023',
 'ZAF/NHLS-UCT-PA-J093/2022',
 'BRA/HRMS_1724/2020',
 'FRA/2212V010351/2022',
 'USA/GA-GPHL-3276/2022',
 'USA/MN-CDC-STM-TZ3G9AY5B/2023',
 'BHR/23410005166_S87_L001/2023',
 'BHR/550701966/2023',
 'USA/ME-CDC-VSX-A061019/2023',
 'BHR/711008191/2022',
 'ZAF/NHLS-UCT-PA-J119/2022',
 'USA/TX-CDC-STM-4V8

In [37]:
metadata = pd.read_csv("./metadata.csv", delimiter="\t")
metadata.head()

strain virus   gisaid_epi_isl genbank_accession  \
0            DOM/UNIBE004/2021  ncov  EPI_ISL_2601037          MZ576192   
1  BGD/icddrb-TND-06-1084/2022  ncov                ?          OQ075389   
2            CHN/RS4_pool/2021  ncov  EPI_ISL_3461148          MZ578003   
3  BGD/icddrb-TND-01-1029/2022  ncov                ?          OQ075393   
4  BGD/icddrb-TND-01-1025/2022  ncov                ?          OQ075384   

  genbank_accession_rev sra_accession        date         region  \
0            MZ576192.1           NaN  2021-06-10  North America   
1            OQ075389.1   SRR22799510  2022-10-01           Asia   
2            MZ578003.1           NaN  2021-01-16           Asia   
3            OQ075393.1   SRR22799544  2022-10-03           Asia   
4            OQ075384.1   SRR22799525  2022-10-01           Asia   

              country            division  ... QC_frame_shifts QC_stop_codons  \
0  Dominican Republic  Dominican Republic  ...            good           good   
1          Bangladesh          Bangladesh  ...            good           good   
2               China               China  ...            good           good   
3          Bangladesh          Bangladesh  ...            good           good   
4          Bangladesh          Bangladesh  ...            good           good   

  QC_overall_score QC_overall_status frame_shifts  \
0        50.173611          mediocre          NaN   
1         6.213018              good          NaN   
2        17.361111              good          NaN   
3         6.213018              good          NaN   
4        21.950960              good          NaN   

                                           deletions insertions  \
0                                        11288-11296        NaN   
1  11288-11296,21633-21641,21992-21994,28362-2837...        NaN   
2                                                NaN        NaN   
3  11288-11296,21633-21641,21992-21994,28362-2837...        NaN   
4  11288-11296,21633-21641,21992-21994,28362-2837...        NaN   

                                       substitutions  \
0  C241T,C1059T,A2941G,C3037T,C8980T,T9867C,A1209...   
1  C241T,A405G,T670G,T1651C,C2790T,G4184A,C4321T,...   
2  C241T,G922A,T1947C,C3037T,G3431T,T5653C,C5812T...   
3  C241T,A405G,T670G,C2790T,G4184A,C4321T,C9344T,...   
4  C401T,A405G,T670G,C2790T,G4184A,C4321T,C9534T,...   

                                     aaSubstitutions  \
0  N:P199L,N:M234I,N:K370N,ORF1a:T265I,ORF1a:L320...   
1  E:T9I,E:T11A,M:Q19E,M:A63T,N:P13L,N:R203K,N:G2...   
2  N:P6S,N:S194L,ORF1a:V561A,ORF1a:V1056L,ORF1a:A...   
3  E:T9I,E:T11A,M:Q19E,M:A63T,N:P13L,N:R203K,N:G2...   
4  E:T9I,E:T11A,M:Q19E,M:A63T,N:P13L,N:R203K,N:G2...   

  clock_deviation,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  
0  -2.0,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,...                                                                                                                                                          
1  -5.0,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,...                                                                                                                                                          
2  3.0,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,...                                                                                                                                                          
3  -8.0,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,...                                                                                                                                                          
4  -25.0,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,...                                                                                                                                                          

[5 rows x 56 columns]

In [35]:
metadata.columns

Index(['strain', 'virus', 'gisaid_epi_isl', 'genbank_accession',
       'genbank_accession_rev', 'sra_accession', 'date', 'region', 'country',
       'division', 'location', 'region_exposure', 'country_exposure',
       'division_exposure', 'segment', 'length', 'host', 'age', 'sex',
       'Nextstrain_clade', 'pango_lineage', 'GISAID_clade', 'originating_lab',
       'submitting_lab', 'authors', 'url', 'title', 'paper_url',
       'date_submitted', 'date_updated', 'sampling_strategy', 'database',
       'Nextclade_pango', 'immune_escape', 'ace2_binding', 'missing_data',
       'divergence', 'nonACGTN', 'coverage', 'rare_mutations',
       'reversion_mutations', 'potential_contaminants', 'QC_missing_data',
       'QC_mixed_sites', 'QC_rare_mutations', 'QC_snp_clusters',
       'QC_frame_shifts', 'QC_stop_codons', 'QC_overall_score',
       'QC_overall_status', 'frame_shifts', 'deletions', 'insertions',
       'substitutions', 'aaSubstitutions',
       'clock_deviation,,,,,,,,,,,,,,,,,,,

In [41]:
i=0
for strain in metadata["strain"]:
    if strain in inf:
        i+=1
i

1107

In [42]:
len(metadata["strain"])

2927

In [43]:
len(inf)

2859

In [44]:
metadata["strain"].value_counts()

DOM/UNIBE004/2021                                      1
TUN/TUN-202129459/2021                                 1
THA/CONI-3932/2022                                     1
Switzerland/SO-ETHZ-38007422/2022                      1
THA/CONI-3970/2023                                     1
                                                      ..
GHA/2230_S4/2020                                       1
LAO/LOMWRU-0841/2022                                   1
KEN/SS11161/2022                                       1
KAZ/03_18/2022                                         1
IMS-11618-CVDP-6922B218-A461-4BC6-A265-4857C0DD6AFA    1
Name: strain, Length: 2927, dtype: int64

In [46]:
metadata[["strain"]].describe()

strain
count                2927
unique               2927
top     DOM/UNIBE004/2021
freq                    1

#### It is interesting to see that we have 2927 unique strains in our metadata but just 1107 of them overlap with 2859 fasta-Sequences.

### MSA

In [59]:
from Bio import AlignIO
from Bio.Align import MultipleSeqAlignment

alignment = AlignIO.read("./aligned.fasta", "fasta")

msa = MultipleSeqAlignment(alignment)

In [61]:
from Bio.Phylo.TreeConstruction import DistanceCalculator 

calculator = DistanceCalculator('identity')

distance_matrix = calculator.get_distance(msa)
print(distance_matrix)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [62]:
from Bio.Phylo.TreeConstruction import DistanceTreeConstructor

constructor = DistanceTreeConstructor()

#UPGMA 
UPGMATree = constructor.upgma(distance_matrix)

In [63]:
import time

print(time.localtime())

time.struct_time(tm_year=2023, tm_mon=3, tm_mday=19, tm_hour=6, tm_min=0, tm_sec=39, tm_wday=6, tm_yday=78, tm_isdst=0)
